tools/yulPhaser/Mutations.cpp

```diff
@@ -0,0 +1,147 @@
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <tools/yulPhaser/Mutations.h>
+
+#include <tools/yulPhaser/SimulationRNG.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <string>
+#include <vector>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::phaser;
+
+function<Mutation> phaser::geneRandomisation(double _chance)
+{
+	return [=](Chromosome const& _chromosome)
+	{
+		vector<string> optimisationSteps;
+		for (auto const& step: _chromosome.optimisationSteps())
+			optimisationSteps.push_back(
+				SimulationRNG::bernoulliTrial(_chance) ?
+				Chromosome::randomOptimisationStep() :
+				step
+			);
+
+		return Chromosome(move(optimisationSteps));
+	};
+}
+
+function<Mutation> phaser::geneDeletion(double _chance)
+{
+	return [=](Chromosome const& _chromosome)
+	{
+		vector<string> optimisationSteps;
+		for (auto const& step: _chromosome.optimisationSteps())
+			if (!SimulationRNG::bernoulliTrial(_chance))
+				optimisationSteps.push_back(step);
+
+		return Chromosome(move(optimisationSteps));
+	};
+}
+
+function<Mutation> phaser::geneAddition(double _chance)
+{
+	return [=](Chromosome const& _chromosome)
+	{
+		vector<string> optimisationSteps;
+
+		if (SimulationRNG::bernoulliTrial(_chance))
+			optimisationSteps.push_back(Chromosome::randomOptimisationStep());
+
+		for (auto const& step: _chromosome.optimisationSteps())
+		{
+			optimisationSteps.push_back(step);
+			if (SimulationRNG::bernoulliTrial(_chance))
+				optimisationSteps.push_back(Chromosome::randomOptimisationStep());
+		}
+
+		return Chromosome(move(optimisationSteps));
+	};
+}
+
+function<Mutation> phaser::alternativeMutations(
+	double _firstMutationChance,
+	function<Mutation> _mutation1,
+	function<Mutation> _mutation2
+)
+{
+	return [=](Chromosome const& _chromosome)
+	{
+		if (SimulationRNG::bernoulliTrial(_firstMutationChance))
+			return _mutation1(_chromosome);
+		else
+			return _mutation2(_chromosome);
+	};
+}
+
+namespace
+{
+
+Chromosome buildChromosomesBySwappingParts(
+	Chromosome const& _chromosome1,
+	Chromosome const& _chromosome2,
+	size_t _crossoverPoint
+)
+{
+	assert(_crossoverPoint <= _chromosome1.length());
+	assert(_crossoverPoint <= _chromosome2.length());
+
+	auto begin1 = _chromosome1.optimisationSteps().begin();
+	auto begin2 = _chromosome2.optimisationSteps().begin();
+
+	return Chromosome(
+		vector<string>(begin1, begin1 + _crossoverPoint) +
```

Comment:
I know this is how it was in the initial paper, but I really think that their chromosomes were much simpler and we should allow for more than one crossover point.

Reply:
This is just the initial algorithm version without anything fancy. I'll include some other crossover variation it in the final PR if my experiments show this one is bad.

As an aside, from what I've read, there's no best crossover operator. Two-point, multi-point, OX, uniform, etc. are good in specific cases but bad in others. One thing that often comes up though is that one-point crossover is usually not the best one :) For example if there are strong dependencies between the initial and the final genes, one-point crossover splits them up and destroys the good properties of the best chromosomes.

In our use case though my intuition is that the closer the optimisation steps are in the sequence the bigger influence they have on each other so one-point crossover should not really break good chromosomes. Maybe another operator would still be better though. For example uniform crossover that goes over both chromosomes and for each gene independently decides whether to swap it or not. I'm testing a few of these operators and I'll report if any makes a difference.